File: code-smells-project/src/middlewares/error_handler.py
"""Centralized error handling — replaces the per-handler try/except that
swallowed exceptions and leaked str(e) to clients."""
from flask import jsonify


class AppError(Exception):
    """Domain/HTTP error raised by controllers and services."""

    def __init__(self, message, status=400):
        super().__init__(message)
        self.message = message
        self.status = status


def register_error_handlers(app):
    @app.errorhandler(AppError)
    def _handle_app_error(err):
        return jsonify({"erro": err.message, "sucesso": False}), err.status

    @app.errorhandler(404)
    def _handle_not_found(_err):
        return jsonify({"erro": "Recurso não encontrado", "sucesso": False}), 404

    @app.errorhandler(Exception)
    def _handle_unexpected(err):
        app.logger.exception("Unhandled error: %s", err)
        return jsonify({"erro": "Erro interno do servidor", "sucesso": False}), 500

File: ecommerce-api-legacy/src/middlewares/errorHandler.js
'use strict';
const { logger } = require('../utils/logger');

// Domain/HTTP error raised by services and controllers.
class AppError extends Error {
  constructor(message, status = 400) {
    super(message);
    this.status = status;
  }
}

// Wraps async route handlers so rejected promises reach the error handler
// instead of crashing or being swallowed.
const asyncHandler = (fn) => (req, res, next) =>
  Promise.resolve(fn(req, res, next)).catch(next);

// Centralized error handler — registered last. Replaces the inconsistent
// per-callback error handling of the original God Class.
function errorHandler(err, _req, res, _next) {
  if (err instanceof AppError) {
    return res.status(err.status).send(err.message);
  }
  logger.error('unhandled', { error: err.message });
  return res.status(500).send('Erro interno do servidor');
}

module.exports = { AppError, asyncHandler, errorHandler };
